File: src/filesystem/docs/tools/remove-page.ts
import type { PageMeta } from '../../page-types';

export const meta: PageMeta = {
  name: 'remove-page',
  description: 'Remove a page entry previously added by md_to_page.py or add_page.py.',
  tech: ['Python'],
  sections: [
    {
      title: 'Overview',
      body: '<p><code>remove_page.py</code> is the inverse of <code>md_to_page.py</code> and <code>add_page.py</code>. It removes the <code>&lt;slug&gt;.ts</code> file from the section and cleans up all references in the section\'s <code>data.ts</code> and <code>index.ts</code>.</p>',
    },
    {
      title: 'Usage',
      body: `<pre><code>python tools/remove_page.py &lt;slug&gt; --path &lt;target-path&gt; [--dry-run]
</code></pre>
<p>Run from the project root.</p>`,
    },
    {
      title: 'Arguments',
      body: `<table>
<thead>
<tr>
<th>Argument</th>
<th>Required</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>slug</code></td>
<td>yes</td>
<td>The page slug — the value of the <code>name</code> field from its frontmatter</td>
</tr>
<tr>
<td><code>--path</code></td>
<td>yes</td>
<td>Directory the page was added under. Use <code>/</code> or <code>~</code> for root (<code>~/</code>). Supports nested paths (e.g. <code>projects</code>, <code>docs/tools</code>).</td>
</tr>
<tr>
<td><code>--dry-run</code></td>
<td>no</td>
<td>Print what would be removed without modifying any files</td>
</tr>
<tr>
<td><code>--tree</code></td>
<td>no</td>
<td>Print the virtual filesystem tree after the operation</td>
</tr>
</tbody>
</table>`,
    },
    {
      title: 'What Gets Removed',
      body: `<p><strong>Section or subdirectory</strong> (<code>--path &lt;section&gt;</code> or <code>--path &lt;section/subdir&gt;</code>):
- <code>src/filesystem/&lt;path&gt;/&lt;slug&gt;.ts</code> — deleted
- <code>src/filesystem/&lt;path&gt;/data.ts</code> — import and array entry removed
- <code>src/filesystem/&lt;path&gt;/index.ts</code> — import and directory entry block removed</p>
<p><strong>Root</strong> (<code>--path /</code>):
- <code>src/filesystem/&lt;slug&gt;.ts</code> — deleted
- <code>src/filesystem/index.ts</code> — import and entry block removed
- <code>src/filesystem/pageRegistry.ts</code> — meta import and entry removed</p>`,
    },
    {
      title: 'Examples',
      body: `<pre><code class="language-bash">python tools/remove_page.py my-project --path projects
python tools/remove_page.py acme-job --path experience --dry-run
python tools/remove_page.py old-entry --path blog --tree

# Root home directory
python tools/remove_page.py readme --path /

# Nested subdirectory
python tools/remove_page.py tree --path docs/tools
</code></pre>`,
    },
    {
      title: 'Notes',
      body: `<ul>
<li>Errors out if <code>&lt;slug&gt;.ts</code> is not found.</li>
<li>Entries not found in <code>data.ts</code> or <code>index.ts</code> are skipped with a warning rather than failing.</li>
<li>After removing, run <code>npm run lint</code> to verify.</li>
</ul>`,
    },
  ],
};

export const readme = `# remove-page

${meta.description}

tech:  ${(meta.tech ?? []).join(' · ')}
`;

File: tools/remove_page.py
#!/usr/bin/env python3
"""
remove_page.py — Remove a page entry previously added by md_to_page.py.

All section directories follow the same pattern, so removal is symmetric:
the entry is removed from <section>/data.ts and <section>/index.ts, and the
.ts file is deleted.

USAGE
    python tools/remove_page.py <slug> --path <target-path> [--dry-run]

EXAMPLES
    python tools/remove_page.py my-project --path projects
    python tools/remove_page.py acme-job --path experience --dry-run
    python tools/remove_page.py tree --path docs/tools
"""

import argparse
import sys
from pathlib import Path

from _lib import (
    ROOT, FILESYSTEM_DIR,
    path_parts, content_path,
    remove_ts_file, remove_from_data_ts, remove_from_index_ts,
    remove_from_root_index_ts, remove_from_root_page_registry,
    print_tree,
)

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        description="Remove a page entry previously added by md_to_page.py.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("slug", help="Page slug (the 'name' field from frontmatter)")
    parser.add_argument("--path", required=True,
                        help="Section/subdirectory the page was added under "
                             "(e.g. 'projects', 'docs/tools')")
    parser.add_argument("--dry-run", action="store_true",
                        help="Print what would be removed without modifying any files")
    parser.add_argument("--tree", action="store_true",
                        help="Print the filesystem tree after the operation")
    args = parser.parse_args()

    slug        = args.slug.strip("/")
    is_root     = args.path.strip() in ("/", "~", "")
    target_path = "/" if is_root else args.path.strip("/")

    print(f"\nRemoving '{slug}' from '{target_path}'" + (" [dry-run]" if args.dry_run else ""))

    if is_root:
        remove_ts_file(slug, FILESYSTEM_DIR, args.dry_run)
        remove_from_root_index_ts(slug, args.dry_run)
        remove_from_root_page_registry(slug, args.dry_run)
    else:
        parts       = path_parts(args.path)
        section_dir = FILESYSTEM_DIR / parts[0]
        cdir        = content_path(section_dir, parts[1:])

        if not cdir.is_dir():
            print(f"ERROR: Directory not found: {cdir.relative_to(ROOT)}")
            sys.exit(1)

        remove_ts_file(slug, cdir, args.dry_run)
        remove_from_data_ts(slug, cdir, args.dry_run)
        remove_from_index_ts(slug, cdir, args.dry_run)

    if not args.dry_run:
        print(f"\nDone. Run 'npm run lint' to verify.")
    else:
        print(f"\n[dry-run complete] No files were modified.")

    if args.tree:
        print()
        print_tree()


if __name__ == "__main__":
    main()
